Integrating graphics with chiventure by displaying a split-screen with a chiventure graphic in the upper half

grkapoor17 opened the pull request:
I worked mostly inside src/ui/src/gui.c trying to integrate a graphical split-screen interface with chiventure games. Particularly, I successfully implemented this split-screen by depicting an output and input text box which relays and processes information respectively related to a particular chiventure game in the lower half of the screen. An image of a chiventure banner is displayed on the upper half of the split-screen. I implemented a run_gui() function, which utilizes a significant amount of code from src/ui/examples/split_screen_example.c. I also had to update the CMakeLists in src/ui to include the raylib library, which enabled me to implement the drawings of images. Moreover, an image of a chiventure banner, "chiventure.png", was added to src/ui as this is the one displayed. Lastly, chiventure.c also had to be slightly updated to include run_gui() as opposed to start_gui(), which was the prior name of the function.

xWngl reviewed (changes requested):
The code itself looks good, but when I was running the GUI, I noticed that closing the GUI using ctrl+d, although it successfully closed the window, resulted in a segfault. It doesn't appear to affect the running of the program at all, but something to take a look at.

grkapoor17 commented:
Following @xWngl 's comment about the segmentation fault, I revisited the code and realized that when the CloseWindow() function was called if ctrl+d is pressed, the code and the loop still runs, i.e, I didn't exit out of the code. Thus, I resolved this problem by exiting out after the CloseWindown() function is called in the conditional checking if the ctrl+d keys are pressed.

xWngl reviewed (approved):
Looks good!

MaxineK36 commented on src/ui/CMakeLists.txt:
```diff
@@ -1,19 +1,25 @@
+    add_subdirectory(examples
+                    EXCLUDE_FROM_ALL)
+else()
+    message(WARNING "Won't build graphics examples because raylib could not be found")
```
Nit: this should now say something like "Won't build graphical interface because raylib could not be found"

MaxineK36 commented on src/ui/src/gui.c:
```diff
@@ -1,45 +1,156 @@
 {
-    int ch;
+    //initialize the window height and width
+    int ScreenWidth = 800;
```
These should be `#define`d constants

MaxineK36 commented on src/ui/src/gui.c:
```diff
@@ -1,45 +1,156 @@
+
+            // Check if more characters have been pressed on the same frame
+            while (key > 0) {
+                // NOTE: Only allow keys in range [32..125]
```
Why only keys in this range?

MaxineK36 replied:
Should also be a block comment

MaxineK36 commented on src/ui/src/gui.c:
```diff
@@ -1,45 +1,156 @@
+            mouseOnText = false;
+
+        if (mouseOnText) {
+        // Get pressed key (character) on the queue
```
nit: indent this comment, and it should be a block comment `/* ... */`

MaxineK36 commented on src/ui/src/gui.c:
```diff
@@ -1,45 +1,156 @@
+        // Get pressed key (character) on the queue
+            int key = GetKeyPressed();
+
+            // Check if more characters have been pressed on the same frame
```
Should also be a block comment

MaxineK36 commented on src/ui/src/gui.c:
```diff
@@ -1,45 +1,156 @@
+                letterCount--;
+                name[letterCount] = '\0';
+
+                if (letterCount < 0) letterCount = 0;
```
For readability, please make this mulitple lines

MaxineK36 commented on src/ui/src/gui.c:
```diff
@@ -1,45 +1,156 @@
+            }
+        }
+
+        if (mouseOnText) framesCounter++;
```
Should also be formatted as multiple lines

MaxineK36 commented on src/ui/src/gui.c:
```diff
@@ -1,45 +1,156 @@
+        else framesCounter = 0;
+        
+        if (IsKeyPressed(KEY_ENTER)) {
+            // use the command to string function to turn name into a command
```
These should all be block comments

MaxineK36 commented on src/ui/src/gui.c:
```diff
@@ -1,45 +1,156 @@
+                letterCount--;
+                name[letterCount] = '\0';
+            }
+	    name[letterCount] = (char) 32;
```
nit: indentation

MaxineK36 commented on src/ui/src/gui.c:
```diff
@@ -1,45 +1,156 @@
+        int heightbuf2 = 150;
+        int rectHeight = 120;
+
+	BeginDrawing();
```
nit: indentation

MaxineK36 commented on src/ui/src/gui.c:
```diff
@@ -1,45 +1,156 @@
+        int pos_x = ScreenWidth/4;
+        int pos_y = ScreenHeight/10;
+        
+        // Loading just one image is a temporary solution
```
Reminder about block comments

MaxineK36 commented on src/ui/src/gui.c:
```diff
@@ -1,45 +1,156 @@
+        DrawRectangle(POS_ZERO, ScreenHeight - heightbuf2, ScreenWidth, rectHeight, WHITE);
+        DrawRectangleLines(POS_ZERO, ScreenHeight - heightbuf2, ScreenWidth, heightbuf2, BLACK);
+
+        int xbuf = 5;
```
Please add some comments explaining this last chunk of code (lines 133 - 151)

MaxineK36 commented on src/ui/src/gui.c:
```diff
@@ -1,45 +1,156 @@
+        Font test = GetFontDefault();
+        DrawTextRec(test, output_text, output, fontSize, fontSpacing, true, BLACK);
+
+    EndDrawing();
```
nit: indentation

MaxineK36 reviewed (changes requested):
A couple of stylistic fixes, but the functionality looks great!

grkapoor17 commented:
Following @MaxineK36 's comments and changes requested, I have added block comments to gui.c explaining what several parts of the code does. I also updated the CMakeLists error message that will load if the raylib library cannot be found. Moreover, I have defined constants in gui.h to remove magic numbers.

MaxineK36 reviewed (approved):
LGTM!

borjasotomayor commented:
Please note that this PR breaks the build in environments that don't have raylib installed (there was a recent post on #general about this). In particular, building the `ui` component _cannot_ depend on raylib being installed, since the UI component only uses a text-based interface that does not depend on Raylib.

xWngl commented:
Modified the chiventure CMake file, ui CMake file, and chiventure.c to only require raylib for run_gui() function. Chiventure can now be built and run without the raylib library, however if you attempt to run a chiventure game with the --gui flag without raylib, it will produce an error message

borjasotomayor reviewed (commented):
The changes to chiventure.c and the CMake files look good to me! (I've also verified that the code builds and runs correctly in environments with and without raylib)

MaxineK36 reviewed (commented):
The updates look great to me!

grkapoor17 commented:
In this pull request, we have successfully merged code from the gui/integrate_graphic into dev. This involved changes to gui.c, gui.h and the CMakeLists. Particularly, we have integrated a graphical interface with chiventure such that if you run a chiventure game with a --gui flag, a graphical split screen will be displayed with the upper half displaying a chiventure banner and the lower half comprises of an input and output text box. This was done using a run_gui() function. Moreover, the image that is displayed, "chiventure.png" was added to src/ui/src and updates also had to be made to the Chiventure CMakeLists as well as the ui CMakeLists. It must be noted that the code was merged twice. This is because there was initially an error that would not let the code compile unless the user had raylib installed and this was problematic as a user does not need raylib to run the normal ui interface. Thus, changes were made to chiventure.c as well as the CMakeLists to fix this issue.

MaxineK36 commented:
Issue Score: ✔️++

Comments:
Excellent work! You did a particularly good job providing a useful closing statement that explains some of the weird PR history. Going forward, you might also want to try using bullet points to make your summary a little easier to read (about 4-5 bullet points is usually equivalent to 1 paragraph for length purposes)

<!-- {"score": "check-plus-plus"} -->
